lib/spack/spack/test/llnl/util/filesystem.py

```diff
@@ -0,0 +1,214 @@
+##############################################################################
+# Copyright (c) 2013-2018, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/spack/spack
+# Please also see the NOTICE and LICENSE files for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+"""Tests for ``llnl/util/filesystem.py``"""
+
+from spack.stage import Stage
+
+import llnl.util.filesystem as fs
+import os
+import pytest
+
+
+@pytest.fixture()
+def stage():
+    """Creates a stage with the directory structure for the tests."""
+
+    s = Stage('copy-install-test')
+    s.create()
+
+    with fs.working_dir(s.path):
+        # Create source file hierarchy
+        fs.touchp('source/1')
+        fs.touchp('source/a/b/2')
+        fs.touchp('source/a/b/3')
+        fs.touchp('source/c/4')
+        fs.touchp('source/c/d/5')
+        fs.touchp('source/c/d/6')
+        fs.touchp('source/c/d/e/7')
+
+        # Create symlink
+        os.symlink(os.path.abspath('source/1'), 'source/2')
+
+        # Create destination directory
+        fs.mkdirp('dest')
+
+    yield s
+
+    s.destroy()
+
+
+class TestCopy:
```

Comment:
Do we have to group these into classes?  I started converting the tests to use straight `pytest` style (mostly free functions + fixtures) instead of `unittest`-style classes (where you use the class to store state with setup and teardown).  You're essentially only using the classes for grouping here -- the tests make no use of `self` -- so can we convert them to regular old functions?  You could convert the class docstrings to comments if you want grouping in the test suite:

```python
#
# Test using a filename as the destination.
#
```

You'd have to make the names unique, though. Does grouping with classes clearly show the class comments when you run `spack test -v`?

Reply:
Grouping with classes is nice because it lets you run all tests for a particular class. It's also nice for logical grouping of similar tests.